### PR TITLE
Suppress the "read-modify-write" error if type is a struct or a class

### PR DESCRIPTION
**Pull request by ghost**
If a struct or class with shared type implements the opUnary or opOpAssign
operator then it is possible to perform "read-modify-write" operations
because the operator's implementation is supposed to perform atomic operations.

The purpose of this modification is to allow the creation of wrappers
(with structs or classes) to run atomic operations silently, for example:

```
shared struct Atomic {
    int a;

    int opUnary(string s)() if (s == "++")
    {
        import core.atomic : atomicOp;
        return atomicOp!"+="(a, 1);
    }
}

Atomic atomicvar;
atomicvar++; // Safe! Atomic struct implements opUnary
```

### CODE REVIEW

File: src/dmd/expression.d [hunk 1]
```diff
@@ -1353,7 +1353,7 @@ extern (C++) abstract class Expression : ASTNode
     extern (D) final bool checkReadModifyWrite(TOK rmwOp, Expression ex = null)
     {
         //printf("Expression::checkReadModifyWrite() %s %s", toChars(), ex ? ex.toChars() : "");
-        if (!type || !type.isShared())
+        if (!type || !type.isShared() || type.isTypeStruct() || type.isTypeClass())
```
Comment (Geod24) [inline src/dmd/expression.d]: That doesn't look right. It's blindly allowing `struct` / `class`

**Review by Geod24 — DISMISSED** [review]
No test

File: src/dmd/expression.d (hunk 1 above)
Comment (ErnyTech) [reply src/dmd/expression.d]: If the class/struct does not implement the operator the compiler will give the error "is not a scalar" (https://github.com/dlang/dmd/blob/d9826c828a8b228713f71136dc6108c42e295352/src/dmd/expressionsem.d#L5574)

So I don't think further checks are necessary

**Comment by ErnyTech** [comment]
Added compilable test

**Comment by thewilsonator** [comment]
```
Searching for tabs
./changelog/pending.dd:321:	int opUnary(string s)() if (s == "++")
./changelog/pending.dd:322:	{
./changelog/pending.dd:323:		import core.atomic : atomicOp;
./changelog/pending.dd:324:		return atomicOp!"+="(a, 1);
./changelog/pending.dd:325:	}
```

File: test/compilable/readmodify_structclass.d [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+		import core.atomic : atomicOp;
+		return atomicOp!"+="(x, 1);
+	}
+}
+
+void main()
+{
+    S s;
+    s++;
+    shared(C) c = new C();
```
Comment (Geod24) [inline test/compilable/readmodify_structclass.d]: Why does this need to be shared if `C` is a `shared class` ?

Comment (ErnyTech) [reply test/compilable/readmodify_structclass.d]: Using shared in the class declaration does not seem to make the class type "implicitly" shared unlike what it does in a struct

**Review by ghost — COMMENTED** [review]
I realized that the formatting has problems (badly converted tabs), I have to solve it

**Comment by ErnyTech** [comment]
> I realized that the formatting has problems (badly converted tabs), I have to solve it

Fixed

**Review by thewilsonator — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**